EIPS/eip-5450.md

```diff
@@ -21,59 +21,102 @@ Currently existing EVM implementations perform a number of validity checks for e
 
-At the same time the following properties are being verified:
+The time and memory complexity if this pass is *O(len(code))*.
 
-1. For each reachable instruction in the section, data stack height is the same for all possible code paths going through this instruction.
-2. For each instruction, data stack always has enough items, i.e. stack underflow is invalid.
-3. For `CALLF` instruction, data stack has enough items to use as input arguments to a called function according to its type defined in the type section.
-4. For every terminating instruction, except `RETF`, data stack is empty after executing it.
-5. For `RETF` instruction, data stack before executing it has exactly `n` items to use as output values returned from a function, where `n` is function's number of outputs according to its type defined in the type section.
-6. Maximum data stack height required by a function does not exceed `1024`.
+#### Operand stack validation
 
-To examine every reachable code path, validation needs to traverse every instruction in order, while also following each non-conditional jump, and following both possible branches for each conditional jump. See below for reference implementation.
+In the second validation phase data-flow (TODO: or control-flow) analysis is performed on the code.
+For each reachable instruction in the code the operand stack height is recorded.
+The first instruction has the recorded stack height equal to the number of inputs to the function matching the code.
 
-The complexity of this traversal is linear in the number of instructions, because each code path is examined only once, and property 1 guarantees no loops in the validation.
+The FIFO queue *worklist* is used for tracking "to be inspected" instructions. Initially, it contains the first instruction.
+
+While *worklist* is not empty, pop an instruction and:
+
+1. Determine the effect the instruction has on the operand stack:
+   1. Check if the recorded stack height satisfies the instruction requirements.
+      (Specifically for `RETF` instruction the recorded stack height must be exactly the number of outputs of the function matching the code).
+   2. Compute new stack height after the instruction execution.
+2. Determine the list of successor instructions than can follow the current instructions:
+   1. The next instruction for all instructions other than terminating instructions and unconditional jump.
+   2. The target of a conditional or unconditional jump.
+3. For each successor instruction:
+   1. Check if the instruction is present in the code (i.e. execution must not "fall off" the code).
+   2. If the instruction does not have stack height recorded (visited for the first time):
+      1. Record the instruction stack height as the value computed in 1.2.
+      2. Add the instruction to the *worklist*.
+   3. Otherwise, check if the recorded stack height equals the value computed in 1.2.
+4. Determine the function maximum operand stack height:
```

Comment:
Maybe better to group this with other stack manipulations of point 1? (make it subpoint of 1?)